From 42bd2c49c49594e7b688f4ae8dc3af59b91b8c78 Mon Sep 17 00:00:00 2001
From: Chalapathi Venkataramashetty <chalapathix.venkataramashetty@intel.com>
Date: Thu, 26 Aug 2021 21:42:07 +0000
Subject: [PATCH] Enable UART mux setting before SOL activation

Switching UART routing when starting obmc-service introduces garbled
character printing out on physical host serial output and it's
inevitable so this commit moves the routing logic into host console
connection flow in IPMI SOL to avoid the issue until SOL is actually
activated.

Tested: SOL activation is working fine using ipmitool

Signed-off-by: Chalapathi Venkataramashetty <chalapathix.venkataramashetty@intel.com>
---
 command/payload_cmds.cpp | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/command/payload_cmds.cpp b/command/payload_cmds.cpp
index b3bcca2..91e7376 100644
--- a/command/payload_cmds.cpp
+++ b/command/payload_cmds.cpp
@@ -6,6 +6,8 @@
 
 #include <ipmid/api.h>
 
+#include <fstream>
+#include <iostream>
 #include <ipmid/api-types.hpp>
 #include <phosphor-logging/lg2.hpp>
 
@@ -15,6 +17,10 @@ namespace sol
 namespace command
 {
 
+static constexpr const char* uartMuxCtrlPath =
+    "/sys/bus/platform/drivers/aspeed-uart-routing/1e789098.uart-routing/hicra";
+static constexpr const char* uartMuxCtrlVal = "0x03450003";
+
 std::vector<uint8_t> activatePayload(const std::vector<uint8_t>& inPayload,
                                      std::shared_ptr<message::Handler>& handler)
 {
@@ -84,6 +90,15 @@ std::vector<uint8_t> activatePayload(const std::vector<uint8_t>& inPayload,
     // Set the current command's socket channel to the session
     handler->setChannelInSession();
 
+    std::ofstream uartMuxFile(uartMuxCtrlPath);
+    if (!uartMuxFile.good())
+    {
+        response->completionCode = IPMI_CC_UNSPECIFIED_ERROR;
+        return outPayload;
+    }
+    uartMuxFile << uartMuxCtrlVal;
+    uartMuxFile.close();
+
     // Start the SOL payload
     try
     {
-- 
2.25.1

